veps/sig-compute/hypervisor-feature-discovery.md

```diff
@@ -0,0 +1,372 @@
+- [ ] (R) Target version is explicitly mentioned and approved
+- [ ] (R) Graduation criteria filled
+
+## Overview
+
+Introduce a centralized capability registry that provides a declarative matrix for expressing feature constraints across hypervisor and architecture combinations. The registry enables validation, test filtering, and clear error messages without duplicating logic across components. 
+
+**Note:** This VEP was extracted from VEP #95 (Hypervisor Abstraction Layer) to allow independent discussion and implementation of the capability registry design. The registry serves as the feature discovery mechanism for the broader hypervisor abstraction framework described in VEP #95. 
+
+## Motivation
+
+- KubeVirt's current architecture-specific validation is scattered across multiple files (`arm64.go`, `s390x.go`), making it difficult to maintain consistency and add new hypervisors.
+- Test filtering logic is duplicated throughout the test suite, requiring manual maintenance when features are added or support changes.
+- Error messages for unsupported features lack context, documentation links, and version information.
+- Adding a new hypervisor requires touching validation logic across multiple components.
+- No clear mechanism to declare when features need explicit validation or blocking without creating scattered conditional code.
+
+## Goals
+
+- Provide a mechanism to declare when features require validation or should be blocked on specific platforms.
+- Enable automatic validation by querying the capability registry instead of scattered conditional logic.
+- Support test filtering so tests only run on compatible infrastructure.
+- Generate rich error messages with explanations, documentation links, and version information when features are unsupported.
+- Keep the registry focused on features that need explicit control—not an exhaustive feature catalog.
+- Make it simple to add validation points and hypervisor-specific constraints.
+
+## Non Goals
+
+- Create an exhaustive catalog of all KubeVirt features (only register capabilities that need validation, blocking, or special metadata).
+- Require registering every feature as a capability (unregistered features are implicitly allowed and handled by normal code paths).
+
+## Definition of Users
+
+- Upstream contributors adding new features or hypervisor support.
+- Platform engineers integrating proprietary hypervisor stacks.
+- Test infrastructure maintainers ensuring tests run only where features are supported.
+- End users receiving clear error messages when requesting unsupported features.
+
+## User Stories
+
+1. As a contributor adding a new hypervisor, I can declare which features need validation or should be blocked, and have validation work automatically without scattered conditional logic.
+2. As a test maintainer, I can decorate tests with capability requirements for features that need platform-specific validation and have them automatically skip on unsupported platforms.
```

Comment:
Tests usually depend on a specific setup, known in advanced (i.e. expected).
Skipping tests based on the avilable platform or setup is not recommended, as that hides unintended mistakes. A test executor, needs to explicitly request the set of tests to run and the setup should support it. If there is no sync between the request and the setup, it should fail.
We have removed skips from the tests codebase for the above reason.

Can you please clarify in which case this is helpful in tests?

Reply:
I agree that "skipping" tests is not the right thing to do. However, if a testcase is testing functionality that is not available on, e.g., ARM64 arch, then that testcase should not be run on the testing lane that targets ARM64 clusters.

For testcase developers, instead of decorating their testcase with labels like `requires-arm64`, they would instead decorate their testcase with `requires-<capability-x>`. The proposed mechanism would determine whether to include or exclude tests with that decorator when running on ARM64 platform.